lib/src/command/add.dart

```diff
@@ -0,0 +1,339 @@
+    var devDependencies = [...original.devDependencies.values];
+
+    if (isDev) {
+      final dependencyNames = dependencies.map((dependency) => dependency.name);
+
+      /// If package is originally in dependencies and we wish to add it to
+      /// dev_dependencies, this is a redundant change, and we should not
+      /// remove the package from dependencies, since it might cause the user's
+      /// code to break.
+      if (dependencyNames.contains(package.name)) {
+        usageException('${package.name} is already in dependencies. '
+            'Please remove existing entry before adding it to dev_dependencies');
+      }
+
+      devDependencies.add(package);
+    } else {
+      final devDependencyNames =
+          devDependencies.map((devDependency) => devDependency.name);
+
+      /// If package is originally in dev_dependencies and we wish to add it to
+      /// dependencies, we remove the package from dev_dependencies, since it is
+      /// now redundant.
+      if (devDependencyNames.contains(package.name)) {
+        log.message('${package.name} was found in dev_dependencies. '
+            'Removing ${package.name} and adding it to dependencies instead.');
+        devDependencies =
+            devDependencies.where((d) => d.name != package.name).toList();
+      }
+
+      dependencies.add(package);
+    }
+
+    return Pubspec(
+      original.name,
+      version: original.version,
+      sdkConstraints: original.sdkConstraints,
+      dependencies: dependencies,
+      devDependencies: devDependencies,
+      dependencyOverrides: original.dependencyOverrides.values,
+    );
+  }
+
+  /// Parse [package] to return the corresponding [PackageRange], as well as its
+  /// representation in `pubspec.yaml`.
+  ///
+  /// [package] must be written in the format
+  /// `<package-name>[:<version-constraint>]`, where quotations should be used
+  /// if necessary.
+  ///
+  /// Examples:
+  /// ```
+  /// retry
+  /// retry:2.0.0
+  /// retry:^2.0.0
+  /// retry:'>=2.0.0'
+  /// retry:'>2.0.0 <3.0.1'
+  /// 'retry:>2.0.0 <3.0.1'
+  /// retry:any
+  /// ```
+  ///
+  /// If a version constraint is provided when the `--path` or any of the
+  /// `--git-<option>` options are used, a [PackageParseError] will be thrown.
+  ///
+  /// Packages must either be a git, hosted, sdk, or path package. Mixing of
+  /// options is not allowed and will cause a [PackageParseError] to be thrown.
+  ///
+  /// If any of the other git options are defined when `--git-url` is not
+  /// defined, an error will be thrown.
+  Pair<PackageRange, dynamic> _parsePackage(String package) {
+    ArgumentError.checkNotNull(package, 'package');
+
+    if ([path != null, hasGitOptions, hasHostOptions, sdk != null]
+            .where((element) => element == true)
+            .length >=
+        2) {
+      usageException(
+          'Packages must either be a git, hosted, sdk, or path package.');
+    }
```

Comment:
nit, it might be nice to print which flags are conflicting.

```dart
final _conflictingFlagSets = [
  ['git-url', 'git-ref', 'git-path'],
  ['hosted-url'],
  ['path'],
  ['sdk'],
];

for (final flag in _conflictingFlagSets.expand((s) => s).where(argResult.wasParsed)) {
  final conflictingFlag = _conflictingFlagSets
    .where((s) => !s.contains(flag))
    .expand((s) => s)
    .firstWhere(argResult.wasParsed, orElse: () => null);
  if (conflictingFlag != null) {
    usageException('Packages can only have one source, pub add flags "--$flag" and "--$conflictingFlag" are conflicting.');
  }
}
```

But I'm sure it can be done smarter than this too :D

Reply:
haha I'm going to steal this :D